pyro/infer/reparam/neutra.py

```diff
@@ -0,0 +1,78 @@
+from torch.distributions import biject_to
+
+import pyro
+import pyro.distributions as dist
+from pyro.distributions.util import sum_rightmost
+from pyro.infer.autoguide.guides import AutoContinuous
+
+from .reparam import Reparam
+
+
+class NeuTraReparam(Reparam):
+    """
+    Neural Transport reparameterizer [1] of multiple latent variables.
+
+    This uses a trained :class:`~pyro.infer.autoguide.AutoContinuous`
+    guide to alter the geometry of a model, typically for use e.g. in MCMC.
+    Example usage::
+
+        # Step 1. Train a guide
+        guide = AutoIAFNormal(model)
+        svi = SVI(model, guide, ...)
+        # ...train the guide...
+
+        # Step 2. Use trained guide in NeuTra MCMC
+        neutra = NeuTraReparam(guide)
+        model = poutine.infer_config(lambda _: {"reparam": neutra})
+        model = poutine.reparam(model)
+        nuts = NUTS(model)
+        # ...now use the model in HMC or NUTS...
+
+    This reparameterization works only for latent variables, not likelihoods.
+    Note that all sites must share a single common :class:`NeuTraReparam`
+    instance, and that the model must have static structure.
+
+    [1] Hoffman, M. et al. (2019)
+        "NeuTra-lizing Bad Geometry in Hamiltonian Monte Carlo Using Neural Transport"
+        https://arxiv.org/abs/1903.03704
+
+    :param ~pyro.infer.autoguide.AutoContinuous guide: A trained guide.
+    """
+    def __init__(self, guide):
+        if not isinstance(guide, AutoContinuous):
+            raise TypeError("NeuTraReparam expected an AutoContinuous guide, but got {}"
+                            .format(type(guide)))
+        self.guide
+        self.x_unconstrained = []
+
+    def __call__(self, name, fn, obs):
+        if name not in self.guide.prototype_trace.nodes:
+            return fn, obs
+        assert obs is None, "NeuTraReparam does not support observe statements"
+        log_density = 0.
+        if not self.x_unconstrained:  # On first sample site.
+            # Sample a shared latent.
+            posterior = self.guide.get_posterior()
+            if not isinstance(posterior, dist.TransformedDistribution):
+                raise ValueError("NeuTraReparam only supports guides whose posteriors are "
+                                 "TransformedDistributions but got a posterior of type {}"
+                                 .format(type(posterior)))
+            t = dist.transforms.ComposeTransform(posterior.transforms)
+            z_unconstrained = pyro.sample("{}_shared_latent".format(name),
+                                          posterior.base_dist.mask(False))
+
+            # Differentiably transform.
+            x_unconstrained = t(z_unconstrained)
+            log_density = t.log_abs_det_jacobian(z_unconstrained, x_unconstrained)
+            self.x_unconstrained = list(reversed(self.guide._unpack_latent(x_unconstrained)))
```

Comment:
I think we'll need to do `list(reversed(list(..)` here.

Reply:
reversed works on iterators

Reply:
That's interesting. It doesn't work with generators though (I guess because it can be an infinite stream). 😞

Reply:
oh sorry. Could  you make the change, taking temporary ownership of this file while you write tests? (sorry this PR is so big, I just wanted to get the interface right, and the best way to do that was to bang on it from every angle at once)

Reply:
Sounds good! I'll handle any minor changes (non-interface related) to get the test working.